File: eclipse/src/com/copilotreview/eclipse/CopilotReviewService.java
package com.copilotreview.eclipse;

import java.io.File;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.eclipse.core.net.proxy.IProxyData;
import org.eclipse.core.net.proxy.IProxyService;
import org.eclipse.core.resources.IFile;
import org.eclipse.core.resources.IMarker;
import org.eclipse.core.resources.IResource;
import org.eclipse.core.runtime.ILog;
import org.eclipse.core.runtime.Platform;
import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceReference;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;

public class CopilotReviewService {

    public static final String MARKER_TYPE = "com.copilotreview.eclipse.reviewMarker";
    private static final ILog LOG = Platform.getLog(CopilotReviewService.class);
    private static CopilotReviewService instance;
    private final Gson gson = new Gson();

    private volatile String cachedToken;
    private volatile long tokenExpiry;

    public static synchronized CopilotReviewService getInstance() {
        if (instance == null) {
            instance = new CopilotReviewService();
        }
        return instance;
    }

    public boolean isGitProject(IFile file) {
        File dir = file.getProject().getLocation().toFile();
        while (dir != null) {
            if (new File(dir, ".git").exists()) return true;
            dir = dir.getParentFile();
        }
        return false;
    }

    public List<ReviewIssue> reviewFile(IFile file) throws Exception {
        String content = Files.readString(file.getLocation().toFile().toPath(), StandardCharsets.UTF_8);
        String fileName = file.getName();
        String lang = file.getFileExtension() != null ? file.getFileExtension() : "unknown";

        List<ReviewIssue> issues = callCopilotForReview(content, fileName, lang);
        applyMarkers(file, issues);
        return issues;
    }

    public void clearMarkers(IResource resource) {
        try {
            resource.deleteMarkers(MARKER_TYPE, true, IResource.DEPTH_INFINITE);
        } catch (Exception e) {
            LOG.warn("Failed to clear markers: " + e.getMessage());
        }
    }

    private void applyMarkers(IFile file, List<ReviewIssue> issues) throws Exception {
        file.deleteMarkers(MARKER_TYPE, true, IResource.DEPTH_ZERO);

        for (ReviewIssue issue : issues) {
            IMarker marker = file.createMarker(MARKER_TYPE);
            marker.setAttribute(IMarker.LINE_NUMBER, issue.line);
            marker.setAttribute(IMarker.MESSAGE, "[Copilot Review] " + issue.message);
            marker.setAttribute(IMarker.SOURCE_ID, "CopilotReview");

            switch (issue.severity) {
                case "error":
                    marker.setAttribute(IMarker.SEVERITY, IMarker.SEVERITY_ERROR);
                    break;
                case "warning":
                    marker.setAttribute(IMarker.SEVERITY, IMarker.SEVERITY_WARNING);
                    break;
                default:
                    marker.setAttribute(IMarker.SEVERITY, IMarker.SEVERITY_INFO);
                    break;
            }
        }
    }

    private List<ReviewIssue> callCopilotForReview(String code, String fileName, String lang) throws Exception {
        String token = getCopilotApiToken();
        String prompt = buildReviewPrompt(code, fileName, lang);

        JsonObject request = new JsonObject();
        JsonArray messages = new JsonArray();

        JsonObject systemMsg = new JsonObject();
        systemMsg.addProperty("role", "system");
        systemMsg.addProperty("content", "You are a code reviewer. You respond only with valid JSON arrays.");
        messages.add(systemMsg);

        JsonObject userMsg = new JsonObject();
        userMsg.addProperty("role", "user");
        userMsg.addProperty("content", prompt);
        messages.add(userMsg);

        request.add("messages", messages);
        request.addProperty("model", "gpt-4o");
        request.addProperty("temperature", 0.1);
        request.addProperty("stream", false);

        URL url = URI.create("https://api.githubcopilot.com/chat/completions").toURL();
        HttpURLConnection conn = openConnection(url);
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Authorization", "Bearer " + token);
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setRequestProperty("Editor-Version", "Eclipse/4.30");
        conn.setRequestProperty("Editor-Plugin-Version", "copilot-code-review/0.0.1");
        conn.setRequestProperty("Openai-Organization", "github-copilot");
        conn.setRequestProperty("Copilot-Integration-Id", "vscode-chat");
        conn.setConnectTimeout(30000);
        conn.setReadTimeout(60000);
        conn.setDoOutput(true);

        byte[] body = gson.toJson(request).getBytes(StandardCharsets.UTF_8);
        try (OutputStream os = conn.getOutputStream()) {
            os.write(body);
        }

        int responseCode = conn.getResponseCode();
        if (responseCode != 200) {
            String errorBody;
            try {
                errorBody = new String(conn.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
            } catch (Exception e) {
                errorBody = "unreadable";
            }
            throw new RuntimeException("Copilot API returned " + responseCode + ": " + errorBody);
        }

        String responseBody = new String(conn.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
        return extractReviewFromResponse(responseBody);
    }

    private String getCopilotApiToken() throws Exception {
        if (cachedToken != null && System.currentTimeMillis() / 1000 < tokenExpiry - 300) {
            return cachedToken;
        }

        String oauthToken = readCopilotOAuthToken();
        if (oauthToken == null) {
            throw new IllegalStateException(
                "Could not find GitHub Copilot OAuth token. Make sure you are signed in to GitHub Copilot.");
        }

        URL url = URI.create("https://api.github.com/copilot_internal/v2/token").toURL();
        HttpURLConnection conn = openConnection(url);
        conn.setRequestMethod("GET");
        conn.setRequestProperty("Authorization", "token " + oauthToken);
        conn.setRequestProperty("Accept", "application/json");
        conn.setRequestProperty("User-Agent", "CopilotCodeReview-Eclipse/0.0.1");
        conn.setConnectTimeout(10000);
        conn.setReadTimeout(10000);

        int responseCode = conn.getResponseCode();
        if (responseCode != 200) {
            String errorBody;
            try {
                errorBody = new String(conn.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
            } catch (Exception e) {
                errorBody = "";
            }
            throw new RuntimeException("Failed to get Copilot token (HTTP " + responseCode + "): " + errorBody);
        }

        String responseBody = new String(conn.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
        JsonObject json = JsonParser.parseString(responseBody).getAsJsonObject();
        cachedToken = json.get("token").getAsString();
        tokenExpiry = json.get("expires_at").getAsLong();

        return cachedToken;
    }

    private String readCopilotOAuthToken() {
        String homeDir = System.getProperty("user.home");
        String[] configPaths = {
            homeDir + "/.config/github-copilot/hosts.json",
            homeDir + "/.config/github-copilot/apps.json",
            homeDir + "/Library/Application Support/github-copilot/hosts.json",
            homeDir + "/Library/Application Support/github-copilot/apps.json",
            homeDir + "/AppData/Local/github-copilot/hosts.json",
            homeDir + "/AppData/Local/github-copilot/apps.json"
        };

        for (String path : configPaths) {
            File file = new File(path);
            if (!file.exists()) continue;

            try {
                String content = Files.readString(file.toPath(), StandardCharsets.UTF_8);
                JsonObject json = JsonParser.parseString(content).getAsJsonObject();

                for (String key : json.keySet()) {
                    JsonElement entry = json.get(key);
                    if (entry != null && entry.isJsonObject()) {
                        JsonObject obj = entry.getAsJsonObject();
                        if (obj.has("oauth_token")) {
                            return obj.get("oauth_token").getAsString();
                        }
                    }
                }

                if (json.has("oauth_token")) {
                    return json.get("oauth_token").getAsString();
                }
            } catch (Exception e) {
                LOG.info("Could not read Copilot config at " + path + ": " + e.getMessage());
            }
        }

        return null;
    }

    private List<ReviewIssue> extractReviewFromResponse(String responseBody) {
        JsonObject json = JsonParser.parseString(responseBody).getAsJsonObject();
        JsonArray choices = json.getAsJsonArray("choices");
        if (choices == null || choices.size() == 0) return List.of();

        JsonObject message = choices.get(0).getAsJsonObject().getAsJsonObject("message");
        String content = message.has("content") ? message.get("content").getAsString() : null;
        if (content == null) return List.of();

        return parseResponse(content);
    }

    List<ReviewIssue> parseResponse(String response) {
        String jsonStr = response.trim();

        // Strip markdown fences
        Pattern fencePattern = Pattern.compile("```(?:json)?\\s*([\\s\\S]*?)```");
        Matcher fenceMatcher = fencePattern.matcher(jsonStr);
        if (fenceMatcher.find()) {
            jsonStr = fenceMatcher.group(1).trim();
        }

        // Find JSON array
        Pattern arrayPattern = Pattern.compile("\\[\\s*[\\s\\S]*]");
        Matcher arrayMatcher = arrayPattern.matcher(jsonStr);
        if (!arrayMatcher.find()) return List.of();
        jsonStr = arrayMatcher.group();

        try {
            List<Map<String, Object>> raw = gson.fromJson(jsonStr,
                    new TypeToken<List<Map<String, Object>>>() {}.getType());
            List<ReviewIssue> issues = new ArrayList<>();
            for (Map<String, Object> map : raw) {
                Number line = (Number) map.get("line");
                String severity = (String) map.get("severity");
                String msg = (String) map.get("message");
                if (line == null || msg == null) continue;
                issues.add(new ReviewIssue(line.intValue(), severity != null ? severity : "warning", msg));
            }
            return issues;
        } catch (Exception e) {
            LOG.warn("Failed to parse review response: " + e.getMessage());
            return List.of();
        }
    }

    private String buildReviewPrompt(String code, String fileName, String lang) {
        return "Review the following " + lang + " file \"" + fileName + "\" for bugs, security issues, performance problems, and code quality concerns.\n\n"
                + "For each issue found, respond with ONLY a JSON array. Each element must have:\n"
                + "- \"line\": the 1-based line number\n"
                + "- \"severity\": one of \"error\", \"warning\", \"info\", \"hint\"\n"
                + "- \"message\": a concise description of the issue\n\n"
                + "If no issues are found, return an empty array: []\n\n"
                + "Do NOT include any text outside the JSON array. No markdown fences.\n\n"
                + "Code:\n" + code;
    }

    private HttpURLConnection openConnection(URL url) throws Exception {
        // Use Eclipse's proxy service if available
        BundleContext context = Activator.getDefault().getBundle().getBundleContext();
        ServiceReference<IProxyService> proxyRef = context.getServiceReference(IProxyService.class);

        if (proxyRef != null) {
            IProxyService proxyService = context.getService(proxyRef);
            if (proxyService != null && proxyService.isProxiesEnabled()) {
                IProxyData[] proxyData = proxyService.select(url.toURI());
                for (IProxyData pd : proxyData) {
                    if (pd.getHost() != null) {
                        LOG.info("Using Eclipse proxy: " + pd.getHost() + ":" + pd.getPort());
                        Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(pd.getHost(), pd.getPort()));
                        return (HttpURLConnection) url.openConnection(proxy);
                    }
                }
            }
        }

        return (HttpURLConnection) url.openConnection();
    }
}

File: intellij/src/main/kotlin/com/copilotreview/CopilotReviewService.kt
package com.copilotreview

import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.google.gson.Gson
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import com.github.copilot.AuthHelper
import java.io.File
import com.intellij.util.net.HttpConfigurable
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import com.intellij.openapi.wm.ToolWindowManager
import java.util.Date
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.runBlocking

class CopilotReviewService(private val project: Project) {

    private val log = Logger.getInstance(CopilotReviewService::class.java)
    private val debounceTimers = ConcurrentHashMap<String, TimerTask>()
    private val timer = Timer("CopilotReviewDebounce", true)
    private val reviewInProgress = ConcurrentHashMap.newKeySet<String>()
    private val gson = Gson()

    @Volatile
    private var cachedToken: String? = null
    @Volatile
    private var tokenExpiry: Long = 0

    var statusCallback: ((String) -> Unit)? = null

    fun isGitProject(): Boolean {
        val basePath = project.basePath ?: return false
        var dir: File? = File(basePath)
        while (dir != null) {
            if (File(dir, ".git").exists()) return true
            dir = dir.parentFile
        }
        return false
    }

    fun isCopilotInstalled(): Boolean {
        val copilotId = PluginId.getId("com.github.copilot")
        val plugin = PluginManagerCore.getPlugin(copilotId)
        return plugin != null && plugin.isEnabled
    }

    fun scheduleReview(file: VirtualFile) {
        val settings = CopilotReviewSettings.getInstance(project).state
        if (!settings.enabled) return

        val excluded = settings.excludedExtensions.split(",").map { it.trim().lowercase() }
        val ext = file.extension?.lowercase() ?: ""
        if (ext in excluded) return

        val path = file.path
        debounceTimers[path]?.cancel()

        val task = object : TimerTask() {
            override fun run() {
                debounceTimers.remove(path)
                reviewFile(file)
            }
        }
        debounceTimers[path] = task
        timer.schedule(task, settings.debounceMs)
    }

    fun reviewFile(file: VirtualFile) {
        val path = file.path
        if (reviewInProgress.contains(path)) return
        reviewInProgress.add(path)

        statusCallback?.invoke("Reviewing...")

        Thread {
            try {
                val (content, lineCount) = ApplicationManager.getApplication().runReadAction<Pair<String, Int>> {
                    val text = String(file.contentsToByteArray())
                    val lines = text.count { it == '\n' } + 1
                    Pair(text, lines)
                }
                val fileName = file.name
                val lang = file.fileType.name

                val issues = callCopilotForReview(content, fileName, lang)

                ApplicationManager.getApplication().invokeLater {
                    ReviewAnnotator.applyAnnotations(project, path, issues, lineCount)
                    statusCallback?.invoke(if (issues.isEmpty()) "No issues" else "${issues.size} issue(s)")

                    // Update the tool window panel
                    val result = ReviewResult(fileName, path, issues, Date())
                    ReviewToolWindowPanel.update(project, result)

                    // Show and activate the tool window
                    val toolWindow = ToolWindowManager.getInstance(project).getToolWindow("Copilot Review")
                    toolWindow?.show()
                }
            } catch (e: Exception) {
                log.warn("Copilot review failed: ${e.message}", e)
                ApplicationManager.getApplication().invokeLater {
                    statusCallback?.invoke("Review failed: ${e.message}")
                }
            } finally {
                reviewInProgress.remove(path)
            }
        }.start()
    }

    private fun callCopilotForReview(code: String, fileName: String, lang: String): List<ReviewIssue> {
        val token = getCopilotApiToken()
        val prompt = buildReviewPrompt(code, fileName, lang)

        val requestBody = gson.toJson(mapOf(
            "messages" to listOf(
                mapOf("role" to "system", "content" to "You are a code reviewer. You respond only with valid JSON arrays."),
                mapOf("role" to "user", "content" to prompt)
            ),
            "model" to "gpt-4o",
            "temperature" to 0.1,
            "stream" to false
        ))

        val url = URI("https://api.githubcopilot.com/chat/completions").toURL()
        val conn = openConnection(url)
        conn.requestMethod = "POST"
        conn.setRequestProperty("Authorization", "Bearer $token")
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty("Editor-Version", "JetBrains-IC/2024.2")
        conn.setRequestProperty("Editor-Plugin-Version", "copilot-code-review/0.0.1")
        conn.setRequestProperty("Openai-Organization", "github-copilot")
        conn.setRequestProperty("Copilot-Integration-Id", "vscode-chat")
        conn.connectTimeout = 30000
        conn.readTimeout = 60000
        conn.doOutput = true

        conn.outputStream.use { it.write(requestBody.toByteArray()) }

        val responseCode = conn.responseCode
        if (responseCode != 200) {
            val errorBody = try {
                conn.errorStream?.bufferedReader()?.readText() ?: "no error body"
            } catch (_: Exception) { "unreadable" }
            throw RuntimeException("Copilot API returned $responseCode: $errorBody")
        }

        val responseBody = conn.inputStream.bufferedReader().readText()
        return extractReviewFromChatResponse(responseBody)
    }

    private fun getCopilotApiToken(): String {
        // Return cached token if still valid (with 5 min buffer)
        if (cachedToken != null && System.currentTimeMillis() / 1000 < tokenExpiry - 300) {
            return cachedToken!!
        }

        // Get OAuth token from the Copilot plugin's auth
        val oauthToken = getOAuthTokenFromCopilotPlugin()
            ?: throw IllegalStateException(
                "Could not get OAuth token from GitHub Copilot. Make sure you are signed in."
            )

        // Exchange OAuth token for a Copilot API session token
        val url = URI("https://api.github.com/copilot_internal/v2/token").toURL()
        val conn = openConnection(url)
        conn.requestMethod = "GET"
        conn.setRequestProperty("Authorization", "token $oauthToken")
        conn.setRequestProperty("Accept", "application/json")
        conn.setRequestProperty("User-Agent", "CopilotCodeReview-IntelliJ/0.0.1")
        conn.connectTimeout = 10000
        conn.readTimeout = 10000

        val responseCode = conn.responseCode
        if (responseCode != 200) {
            val errorBody = try {
                conn.errorStream?.bufferedReader()?.readText() ?: ""
            } catch (_: Exception) { "" }
            throw RuntimeException("Failed to get Copilot token (HTTP $responseCode): $errorBody")
        }

        val responseBody = conn.inputStream.bufferedReader().readText()
        val json = JsonParser.parseString(responseBody).asJsonObject
        cachedToken = json.get("token").asString
        tokenExpiry = json.get("expires_at").asLong

        return cachedToken!!
    }

    private fun getOAuthTokenFromCopilotPlugin(): String? {
        return try {
            val accounts: Set<com.github.copilot.GitHubAccountCredentials> = runBlocking { AuthHelper.getAccounts() }
            accounts.firstOrNull()?.token
        } catch (e: Exception) {
            log.warn("[CopilotReview] Failed to get token from Copilot plugin: ${e.message}")
            null
        }
    }

    private fun extractReviewFromChatResponse(responseBody: String): List<ReviewIssue> {
        val json = JsonParser.parseString(responseBody).asJsonObject
        val choices = json.getAsJsonArray("choices")
        if (choices == null || choices.size() == 0) return emptyList()

        val message = choices[0].asJsonObject.getAsJsonObject("message")
        val content = message.get("content")?.asString ?: return emptyList()

        return parseResponse(content)
    }

    private fun buildReviewPrompt(code: String, fileName: String, lang: String): String {
        return """Review the following $lang file "$fileName" for bugs, security issues, performance problems, and code quality concerns.

For each issue found, respond with ONLY a JSON array. Each element must have:
- "line": the 1-based line number
- "severity": one of "error", "warning", "info", "hint"
- "message": a concise description of the issue

If no issues are found, return an empty array: []

Do NOT include any text outside the JSON array. No markdown fences.

Code:
$code"""
    }

    fun parseResponse(response: String): List<ReviewIssue> {
        var jsonStr = response.trim()

        // Strip markdown fences
        val fenceMatch = Regex("```(?:json)?\\s*([\\s\\S]*?)```").find(jsonStr)
        if (fenceMatch != null) {
            jsonStr = fenceMatch.groupValues[1].trim()
        }

        // Find JSON array
        val arrayMatch = Regex("\\[\\s*[\\s\\S]*]").find(jsonStr) ?: return emptyList()
        jsonStr = arrayMatch.value

        return try {
            val type = object : TypeToken<List<Map<String, Any>>>() {}.type
            val raw: List<Map<String, Any>> = gson.fromJson(jsonStr, type)
            raw.mapNotNull { map ->
                val line = (map["line"] as? Number)?.toInt() ?: return@mapNotNull null
                val severity = map["severity"] as? String ?: "warning"
                val message = map["message"] as? String ?: return@mapNotNull null
                ReviewIssue(line, severity, message)
            }
        } catch (e: Exception) {
            log.warn("Failed to parse review response: ${e.message}")
            emptyList()
        }
    }

    private fun openConnection(url: URL): HttpURLConnection {
        return HttpConfigurable.getInstance().openHttpConnection(url.toString()) as HttpURLConnection
    }

    fun dispose() {
        timer.cancel()
        debounceTimers.clear()
    }

    companion object {
        fun getInstance(project: Project): CopilotReviewService {
            return project.getService(CopilotReviewService::class.java)
        }
    }
}
